File: src/services/ReportService.cpp
#include "ReportService.h"

#include "domain/AssessmentStatus.h"
#include "domain/Baustein.h"
#include "domain/Requirement.h"
#include "domain/ProtectionNeed.h"
#include "domain/RequirementLevel.h"
#include "domain/Standard.h"
#include "domain/TargetObject.h"

#include <QDate>
#include <QCoreApplication>
#include <QHash>
#include <algorithm>

namespace {

QString reportServiceTr(const char *text)
{
    return QCoreApplication::translate("ReportService", text);
}

} // namespace

ReportService::ReportService(ICatalogRepository &catalog,
                             IProjectRepository &project,
                             ITargetObjectRepository &targetObjects,
                             IMeasureRepository &measures)
    : m_catalog(catalog)
    , m_project(project)
    , m_targetObjects(targetObjects)
    , m_measures(measures)
{
}

QList<ReportRow> ReportService::buildSollIstReport(int projectId,
                                                   int targetObjectId,
                                                   const QString &catalogVersion) const
{
    QList<ReportRow> rows;

    const QList<Baustein> allBausteine =
        m_catalog.loadBausteine(StandardType::ITGrundschutz, catalogVersion);
    QHash<int, Baustein> bausteinById;
    for (const Baustein &baustein : allBausteine)
        bausteinById.insert(baustein.id, baustein);

    const QList<TargetObject> targetObjects = m_targetObjects.loadTargetObjects(projectId);
    for (const TargetObject &targetObject : targetObjects) {
        if (targetObjectId != 0 && targetObject.id != targetObjectId)
            continue;

        const QHash<int, ApplicabilityStatus> applicabilityMap =
            m_targetObjects.loadApplicabilityMap(projectId, targetObject.id);
        if (applicabilityMap.isEmpty())
            continue;

        const QHash<int, int> measureCounts = m_measures.measureCounts(projectId, targetObject.id);

        for (auto it = applicabilityMap.constBegin(); it != applicabilityMap.constEnd(); ++it) {
            if (it.value() != ApplicabilityStatus::Required
                && it.value() != ApplicabilityStatus::Possible) {
                continue;
            }

            const QList<Requirement> requirements = m_catalog.loadRequirements(it.key());
            for (const Requirement &requirement : requirements) {
                if (requirement.withdrawn)
                    continue;
                if (!requirementLevelApplies(requirement.level, targetObject.protectionNeed))
                    continue;

                ReportRow row;
                row.targetObjectId = targetObject.id;
                row.targetObjectName = targetObject.name;
                row.bausteinExternalId = requirement.bausteinExternalId;
                row.requirementExternalId = requirement.externalId;
                row.requirementTitle = requirement.title;
                row.level = requirementLevelToString(requirement.level);
                row.applicability = it.value();
                row.bausteinTitle = bausteinById.value(it.key()).title;

                const RequirementAssessment assessment = m_project.loadAssessment(
                    projectId, targetObject.id, requirement.id);
                row.status = assessment.status;
                row.responsible = assessment.responsible;
                row.dueDate = assessment.dueDate;
                row.measureCount = measureCounts.value(requirement.id, 0);
                row.overdue = assessment.dueDate.isValid()
                              && assessment.dueDate < QDate::currentDate()
                              && assessment.status != AssessmentStatus::Fulfilled
                              && assessment.status != AssessmentStatus::NotApplicable;

                rows.append(row);
            }
        }
    }

    std::sort(rows.begin(), rows.end(), [](const ReportRow &a, const ReportRow &b) {
        if (a.targetObjectName != b.targetObjectName)
            return a.targetObjectName < b.targetObjectName;
        if (a.bausteinExternalId != b.bausteinExternalId)
            return a.bausteinExternalId < b.bausteinExternalId;
        return a.requirementExternalId < b.requirementExternalId;
    });

    return rows;
}

ReportSummary ReportService::summarize(const QList<ReportRow> &rows) const
{
    ReportSummary summary;
    summary.totalRequirements = rows.size();

    for (const ReportRow &row : rows) {
        switch (row.status) {
        case AssessmentStatus::Open:
            ++summary.openCount;
            break;
        case AssessmentStatus::Partial:
            ++summary.partialCount;
            break;
        case AssessmentStatus::Fulfilled:
            ++summary.fulfilledCount;
            break;
        case AssessmentStatus::NotApplicable:
            ++summary.notApplicableCount;
            break;
        }
        if (row.overdue)
            ++summary.overdueCount;
        summary.measureCount += row.measureCount;
    }

    return summary;
}

QHash<int, ReportSummary> ReportService::summarizeByTargetObject(int projectId,
                                                                 const QString &catalogVersion) const
{
    return summarizeByTargetObject(buildSollIstReport(projectId, 0, catalogVersion));
}

QHash<int, ReportSummary> ReportService::summarizeByTargetObject(const QList<ReportRow> &rows) const
{
    QHash<int, QList<ReportRow>> grouped;
    for (const ReportRow &row : rows)
        grouped[row.targetObjectId].append(row);

    QHash<int, ReportSummary> summaries;
    for (auto it = grouped.constBegin(); it != grouped.constEnd(); ++it)
        summaries.insert(it.key(), summarize(it.value()));
    return summaries;
}

int ReportService::progressPercent(const ReportSummary &summary)
{
    return reportProgressPercent(summary);
}

QString ReportService::formatSummaryText(const ReportSummary &summary)
{
    if (summary.totalRequirements == 0)
        return reportServiceTr("Keine anwendbaren Anforderungen");

    const int percent = progressPercent(summary);
    QString text = reportServiceTr("%1% abgeschlossen (%2/%3 Anforderungen)")
                       .arg(percent)
                       .arg(summary.fulfilledCount + summary.notApplicableCount)
                       .arg(summary.totalRequirements);

    QStringList details;
    if (summary.openCount > 0)
        details << reportServiceTr("%1 offen").arg(summary.openCount);
    if (summary.partialCount > 0)
        details << reportServiceTr("%1 teilweise").arg(summary.partialCount);
    if (summary.overdueCount > 0)
        details << reportServiceTr("%1 überfällig").arg(summary.overdueCount);
    if (summary.measureCount > 0)
        details << reportServiceTr("%1 Maßnahmen").arg(summary.measureCount);

    if (!details.isEmpty())
        text += QStringLiteral(" — ") + details.join(QStringLiteral(", "));

    return text;
}

QString ReportService::formatTreeProgressSuffix(const ReportSummary &summary)
{
    if (summary.totalRequirements == 0)
        return {};

    QString suffix = QStringLiteral(" · %1%").arg(progressPercent(summary));
    if (summary.overdueCount > 0)
        suffix += reportServiceTr(", %1 überfällig").arg(summary.overdueCount);
    return suffix;
}

File: Delphi/Services/ReportService.pas
unit ReportService;

interface

uses
  System.SysUtils, System.Generics.Collections, System.Math, IsmsDomain,
  RepositoryBase;

type
  TReportService = class
  private
    FCatalog: TCatalogRepositoryBase;
    FProject: TProjectRepositoryBase;
    FTarget: TTargetObjectRepositoryBase;
    FMeasure: TMeasureRepositoryBase;
    class procedure SortRows(var ARows: TArray<TReportRow>);
  public
    constructor Create(ACatalog: TCatalogRepositoryBase; AProject: TProjectRepositoryBase;
      ATarget: TTargetObjectRepositoryBase; AMeasure: TMeasureRepositoryBase);
    function BuildSollIstReport(AProjectId, ATargetObjectId: Integer;
      const ACatalogVersion: string): TArray<TReportRow>;
    function Summarize(const ARows: TArray<TReportRow>): TReportSummary;
    function SummarizeByTargetObject(const ARows: TArray<TReportRow>): TDictionary<Integer, TReportSummary>;
    function SummarizeByTargetObjectForProject(AProjectId: Integer;
      const ACatalogVersion: string): TDictionary<Integer, TReportSummary>;
    class function ProgressPercent(const ASummary: TReportSummary): Integer;
    class function FormatSummaryText(const ASummary: TReportSummary): string;
    class function FormatTreeProgressSuffix(const ASummary: TReportSummary): string;
  end;

implementation

constructor TReportService.Create(ACatalog: TCatalogRepositoryBase; AProject: TProjectRepositoryBase;
  ATarget: TTargetObjectRepositoryBase; AMeasure: TMeasureRepositoryBase);
begin
  inherited Create;
  FCatalog := ACatalog;
  FProject := AProject;
  FTarget := ATarget;
  FMeasure := AMeasure;
end;

class procedure TReportService.SortRows(var ARows: TArray<TReportRow>);
var
  I, J: Integer;
  Temp: TReportRow;
begin
  for I := 0 to High(ARows) - 1 do
    for J := I + 1 to High(ARows) do
    begin
      if (ARows[J].TargetObjectName < ARows[I].TargetObjectName) or
         ((ARows[J].TargetObjectName = ARows[I].TargetObjectName) and
          (ARows[J].BausteinExternalId < ARows[I].BausteinExternalId)) or
         ((ARows[J].TargetObjectName = ARows[I].TargetObjectName) and
          (ARows[J].BausteinExternalId = ARows[I].BausteinExternalId) and
          (ARows[J].RequirementExternalId < ARows[I].RequirementExternalId)) then
      begin
        Temp := ARows[I];
        ARows[I] := ARows[J];
        ARows[J] := Temp;
      end;
    end;
end;

function TReportService.BuildSollIstReport(AProjectId, ATargetObjectId: Integer;
  const ACatalogVersion: string): TArray<TReportRow>;
var
  Rows: TArray<TReportRow>;
  AllBausteine: TArray<TBaustein>;
  BausteinById: TDictionary<Integer, TBaustein>;
  TargetObjects: TArray<TTargetObject>;
  TargetObject: TTargetObject;
  ApplicabilityMap: TDictionary<Integer, TApplicabilityStatus>;
  MeasureCounts: TDictionary<Integer, Integer>;
  Baustein: TBaustein;
  Pair: TPair<Integer, TApplicabilityStatus>;
  Requirements: TArray<TRequirement>;
  Requirement: TRequirement;
  Assessment: TRequirementAssessment;
  Row: TReportRow;
  MeasureCnt: Integer;
begin
  SetLength(Rows, 0);
  AllBausteine := FCatalog.LoadBausteine(stITGrundschutz, ACatalogVersion);
  BausteinById := TDictionary<Integer, TBaustein>.Create;
  try
    for Baustein in AllBausteine do
      BausteinById.Add(Baustein.Id, Baustein);

    TargetObjects := FTarget.LoadTargetObjects(AProjectId);
    for TargetObject in TargetObjects do
    begin
      if (ATargetObjectId <> 0) and (TargetObject.Id <> ATargetObjectId) then
        Continue;

      ApplicabilityMap := FTarget.LoadApplicabilityMap(AProjectId, TargetObject.Id);
      try
        if ApplicabilityMap.Count = 0 then
          Continue;

        MeasureCounts := FMeasure.MeasureCounts(AProjectId, TargetObject.Id);
        try
          for Pair in ApplicabilityMap do
          begin
            if (Pair.Value <> apRequired) and (Pair.Value <> apPossible) then
              Continue;
            if not BausteinById.ContainsKey(Pair.Key) then
              Continue;

            Requirements := FCatalog.LoadRequirements(Pair.Key);
            for Requirement in Requirements do
            begin
              if Requirement.Withdrawn then
                Continue;
              if not RequirementLevelApplies(Requirement.Level, TargetObject.ProtectionNeed) then
                Continue;

              FillChar(Row, SizeOf(Row), 0);
              Row.TargetObjectId := TargetObject.Id;
              Row.TargetObjectName := TargetObject.Name;
              Row.BausteinExternalId := Requirement.BausteinExternalId;
              Row.BausteinTitle := BausteinById[Pair.Key].Title;
              Row.RequirementExternalId := Requirement.ExternalId;
              Row.RequirementTitle := Requirement.Title;
              Row.Level := RequirementLevelToString(Requirement.Level);
              Row.Applicability := Pair.Value;

              Assessment := FProject.LoadAssessment(AProjectId, TargetObject.Id, Requirement.Id);
              Row.Status := Assessment.Status;
              Row.Responsible := Assessment.Responsible;
              Row.DueDate := Assessment.DueDate;
              if MeasureCounts.TryGetValue(Requirement.Id, MeasureCnt) then
                Row.MeasureCount := MeasureCnt;
              Row.Overdue := IsValidDate(Assessment.DueDate) and
                (Assessment.DueDate < Date) and
                (Assessment.Status <> asFulfilled) and
                (Assessment.Status <> asNotApplicable);

              SetLength(Rows, Length(Rows) + 1);
              Rows[High(Rows)] := Row;
            end;
          end;
        finally
          MeasureCounts.Free;
        end;
      finally
        ApplicabilityMap.Free;
      end;
    end;
  finally
    BausteinById.Free;
  end;
  SortRows(Rows);
  Result := Rows;
end;

function TReportService.Summarize(const ARows: TArray<TReportRow>): TReportSummary;
var
  Row: TReportRow;
begin
  FillChar(Result, SizeOf(Result), 0);
  Result.TotalRequirements := Length(ARows);
  for Row in ARows do
  begin
    case Row.Status of
      asOpen: Inc(Result.OpenCount);
      asPartial: Inc(Result.PartialCount);
      asFulfilled: Inc(Result.FulfilledCount);
      asNotApplicable: Inc(Result.NotApplicableCount);
    end;
    if Row.Overdue then
      Inc(Result.OverdueCount);
    Inc(Result.MeasureCount, Row.MeasureCount);
  end;
end;

function TReportService.SummarizeByTargetObject(const ARows: TArray<TReportRow>): TDictionary<Integer, TReportSummary>;
var
  Grouped: TDictionary<Integer, TList<TReportRow>>;
  Row: TReportRow;
  List: TList<TReportRow>;
  Pair: TPair<Integer, TList<TReportRow>>;
  RowArray: TArray<TReportRow>;
begin
  Result := TDictionary<Integer, TReportSummary>.Create;
  Grouped := TDictionary<Integer, TList<TReportRow>>.Create;
  try
    for Row in ARows do
    begin
      if not Grouped.TryGetValue(Row.TargetObjectId, List) then
      begin
        List := TList<TReportRow>.Create;
        Grouped.Add(Row.TargetObjectId, List);
      end;
      List.Add(Row);
    end;
    for Pair in Grouped do
    begin
      RowArray := Pair.Value.ToArray;
      Result.Add(Pair.Key, Summarize(RowArray));
    end;
  finally
    for Pair in Grouped do
      Pair.Value.Free;
    Grouped.Free;
  end;
end;

function TReportService.SummarizeByTargetObjectForProject(AProjectId: Integer;
  const ACatalogVersion: string): TDictionary<Integer, TReportSummary>;
begin
  Result := SummarizeByTargetObject(BuildSollIstReport(AProjectId, 0, ACatalogVersion));
end;

class function TReportService.ProgressPercent(const ASummary: TReportSummary): Integer;
begin
  Result := ReportProgressPercent(ASummary);
end;

class function TReportService.FormatSummaryText(const ASummary: TReportSummary): string;
var
  Details: TArray<string>;
  DetailCount: Integer;

  procedure AddDetail(const AText: string);
  begin
    SetLength(Details, DetailCount + 1);
    Details[DetailCount] := AText;
    Inc(DetailCount);
  end;

begin
  if ASummary.TotalRequirements = 0 then
    Exit('Keine anwendbaren Anforderungen');

  Result := Format('%d%% abgeschlossen (%d/%d Anforderungen)',
    [ProgressPercent(ASummary),
     ASummary.FulfilledCount + ASummary.NotApplicableCount,
     ASummary.TotalRequirements]);

  DetailCount := 0;
  if ASummary.OpenCount > 0 then
    AddDetail(Format('%d offen', [ASummary.OpenCount]));
  if ASummary.PartialCount > 0 then
    AddDetail(Format('%d teilweise', [ASummary.PartialCount]));
  if ASummary.OverdueCount > 0 then
    AddDetail(Format('%d überfällig', [ASummary.OverdueCount]));
  if ASummary.MeasureCount > 0 then
    AddDetail(Format('%d Maßnahmen', [ASummary.MeasureCount]));

  if DetailCount > 0 then
    Result := Result + ' — ' + string.Join(', ', Details);
end;

class function TReportService.FormatTreeProgressSuffix(const ASummary: TReportSummary): string;
begin
  Result := '';
  if ASummary.TotalRequirements = 0 then
    Exit;
  Result := Format(' · %d%%', [ProgressPercent(ASummary)]);
  if ASummary.OverdueCount > 0 then
    Result := Result + Format(', %d überfällig', [ASummary.OverdueCount]);
end;

end.
